nextcloud_odoo_sync/models/calendar_event.py

```diff
@@ -1,125 +1,256 @@
-            vals['nc_synced'] = False
+        """
+        Inherited odoo base function
+        :params vals: Dictionary of record changes
+        :return Super: add changes into this predefined functions
+        """
+        if not self._context.get("sync", False) and "nc_synced" not in vals:
+            vals["nc_synced"] = False
+        for record in self:
+            # Detach the record from recurring event whenever an edit was made
+            # to make it compatible when synced to Nextcloud calendar
+            if record.recurrence_id:
+                detach = False
+                fields_to_update = list(vals.keys())
+                ex_fields = [
+                    "nc_uid",
+                    "nc_rid",
+                    "nc_hash_ids",
+                    "nc_synced",
+                    "nc_to_delete",
+                    "recurrence_id",
+                    "nc_calendar_select",
+                ]
+                for f in fields_to_update:
+                    if f not in ex_fields:
+                        detach = True
+                        break
+                if detach:
+                    vals.update({"nc_detach": True})
         return super(CalendarEvent, self).write(vals)
 
     def unlink(self):
+        """
+        We can"t delete an event that is also in Nextcloud Calendar.
+        Otherwise we would have no clue that the event must must deleted
+        from Nextcloud Calendar at the next sync. We just mark the event as to
+        delete (nc_to_delete=True) before we sync.
+        """
         for record in self:
-            if not self._context.get('sync', False) and record.nc_require_calendar and not self.env.context.get('force_delete', False):
-                record.write({'nc_to_delete': True})
+            if record.nc_uid and not self._context.get("force_delete", False):
+                record.write({"nc_to_delete": True})
             else:
+                if record.recurrence_id:
+                    nc_exdates = (
+                        ast.literal_eval(str(record.recurrence_id.nc_exdate))
+                        if record.recurrence_id.nc_exdate
+                        else []
+                    )
+                    start_date = record.start.strftime("%Y%m%dT%H%M%S")
+                    if record.allday:
+                        start_date = record.start_date.strftime("%Y%m%d")
+                    nc_exdates.append(start_date)
+                    record.recurrence_id.write({"nc_exdate": nc_exdates})
                 return super(CalendarEvent, self).unlink()
+
+
+class CalendarEventNchash(models.Model):
+    _name = "calendar.event.nchash"
+    _description = "Calendar Event Nextcloud Hash"
+
+    calendar_event_id = fields.Many2one(
+        "calendar.event", "Calendar Event", ondelete="cascade"
+    )
+    user_id = fields.Many2one(
+        "res.users", "Odoo User", related="nc_sync_user_id.user_id", store=True
+    )
```

Comment:
`user_id` is a related field, so no need to specify the comodel_name

```suggestion
    user_id = fields.Many2one(related="nc_sync_user_id.user_id", store=True)
```